btb/hyper_parameter.py

```diff
@@ -1,48 +1,178 @@
 from builtins import object, str as newstr
-from collections import namedtuple
+from collections import namedtuple, defaultdict
+import random
+import numpy as np
+import math
+import operator
 
-class ParamTypes(object):
-	INT = "int"
-	INT_EXP = "int_exp"
-	INT_CAT = "int_cat"
-	FLOAT = "float"
-	FLOAT_EXP = "float_exp"
-	FLOAT_CAT = "float_cat"
-	STRING = "string"
-	BOOL = "bool"
-
-# List of exponential hyperparameter types
-EXP_TYPES = [ParamTypes.INT_EXP, ParamTypes.FLOAT_EXP]
 
-# List of categorical hyperparameter types
-CAT_TYPES = [ParamTypes.INT_CAT, ParamTypes.FLOAT_CAT, ParamTypes.STRING,
-             ParamTypes.BOOL]
+class ParamTypes(object):
+    INT = "int"
+    INT_EXP = "int_exp"
+    INT_CAT = "int_cat"
+    FLOAT = "float"
+    FLOAT_EXP = "float_exp"
+    FLOAT_CAT = "float_cat"
+    STRING = "string"
+    BOOL = "bool"
 
 
-# our HyperParameter object
+# HyperParameter object
 class HyperParameter(object):
-    def __init__(self, typ, rang):
+    def __new__(cls, typ, rang):
+        class_generator = {
```

Comment:
Since this is always the same, it should be defined at module level as a constant (`CLASS_GENERATOR`) instead of here.
This way, we prevent creating an instance of the dict at every call.

Reply:
I don't believe it is possible to define it at the module level, as the classes are defined later in the file. Having it at the module level leads to name errors like `NameError: name 'IntHyperParameter' is not defined`

Reply:
@lauragustafson For it to work, the variable definition has to be after the class definitions, so I suggest putting it at the end of the file.